lightningd/channel.c

```diff
@@ -9,9 +9,37 @@
 #include <lightningd/peer_control.h>
 #include <lightningd/subd.h>
 
+void channel_set_owner(struct channel *channel, struct subd *owner)
+{
+	struct subd *old_owner = channel->owner;
+	channel->owner = owner;
+
+	if (old_owner)
+		subd_release_peer(old_owner, channel2peer(channel));
+}
+
 static void destroy_channel(struct channel *channel)
 {
+	/* Free any old owner still hanging around. */
+	channel_set_owner(channel, NULL);
+
 	list_del_from(&channel->peer->channels, &channel->list);
+
+	/* Last one out frees the peer */
+	if (list_empty(&channel->peer->channels))
+		tal_free(channel->peer);
+}
+
+/* This lets us give a more detailed error than just a destructor. */
+void free_channel(struct channel *channel, const char *why)
+{
+	if (channel->opening_cmd) {
+		command_fail(channel->opening_cmd, "%s", why);
+		channel->opening_cmd = NULL;
+	}
+	wallet_channel_delete(channel->peer->ld->wallet, channel->dbid,
```

Comment:
Here we again use a destructor to actually delete stuff in the DB. This is bad IMHO since the lifetime in memory is not the same as the lifetime in the DB. If we do a clean shutdown we may end up deleting all channels from the DB...

I have a strong preference for explicit DB operations rather than hiding them in destructors.

Reply:
That's not the destructor. The destructors are all called destroy_xxx for that reason.

TBH I wanted the destructor-deletes semantic, but as you point out, shutdown would clean our db:)

Reply:
>The destructors are all called destroy_xxx for that reason.

This surprises me.  There are a few destructors that are not named by this convention (`remove_timer`, `invoice_waiter_dtor`, `free_subd_req`, `remove_unstored_payment`). How strict is this rule, and should we have a cleanup issue for this?

Reply:
Ok, then I propose renaming `free_channel` to `delete_peer`, since `free_*` suggests a memory operation to me :-)

Reply:
Agreed, will push on top.

And will push destroy renaming too...